x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicensedAllocatedPersistentTask.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.license;
+
+import org.elasticsearch.persistent.AllocatedPersistentTask;
+import org.elasticsearch.persistent.PersistentTasksService;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.tasks.TaskManager;
+
+import java.util.Map;
+
+/**
+ * An AllocatedPersistentTask which automatically tracks as a licensed feature usage.
+ */
+public class LicensedAllocatedPersistentTask extends AllocatedPersistentTask {
+    private final LicensedFeature.Persistent licensedFeature;
+    private final String featureContext;
+    private final XPackLicenseState licenseState;
+
+    public LicensedAllocatedPersistentTask(long id, String type, String action, String description, TaskId parentTask,
+                                           Map<String, String> headers, LicensedFeature.Persistent feature, String featureContext,
+                                           XPackLicenseState licenseState) {
+        super(id, type, action, description, parentTask, headers);
+        this.licensedFeature = feature;
+        this.featureContext = featureContext;
+        this.licenseState = licenseState;
+        licensedFeature.startTracking(licenseState, featureContext);
```

Comment:
Looking at the taskManager code, I think there is a small chance that a task object is destroyed but none of the cancellations methods are called. 

https://github.com/elastic/elasticsearch/blob/7d980ad6e58e0c8e65aed86c5912dc18f0efa8e7/server/src/main/java/org/elasticsearch/persistent/PersistentTasksNodeService.java#L178-L202

Note if `init` is called and throws, only `taskManager.unregister(task);` is called. I don't think this calls any of these internal life time tracking methods. 

I am thinking that `licensedFeature.startTracking(licenseState, featureContext);` should probably be called in `init` to avoid this weird condition.

What do you think?

Reply:
It seems like init is being abused currently. It's purpose, IMO is to just set some references, not to do any work? But at least rollup seems to utilize this.

What if instead I were to make init public and final? Public is so that tests can still call it, and then final so that there is no chance some extra work is done that could actually throw an exception.

Reply:
@rjernst if we can make that guarantee of it not throwing, it makes sense to me. All it's doing is setting internal values...so I don't know why it would ever throw. 

So, I think fix init, or somehow make sure that `stopTracking` is called if `init` throws in this execution path.